service/history/workflow/mutable_state_impl.go

```diff
@@ -5367,6 +5411,64 @@ func (ms *MutableStateImpl) closeTransactionUpdateTransitionHistory(
 	return nil
 }
 
+func (ms *MutableStateImpl) closeTransactionTrackLastUpdateVersionedTransition(
+	transactionPolicy TransactionPolicy,
+) {
+	if transactionPolicy != TransactionPolicyActive {
+		// TODO: replication/standby logic will need a different way for updating LastUpdatedVersionedTransition
+		// when reapplying history, especially when history replication tasks got batched.
+		return
+	}
+
+	if len(ms.executionInfo.TransitionHistory) == 0 {
+		// transition history is not enabled
+		return
+	}
+
+	transitionHistory := ms.executionInfo.TransitionHistory
+	currentVersionedTransition := transitionHistory[len(transitionHistory)-1]
+	for _, activityInfo := range ms.updateActivityInfos {
+		activityInfo.LastUpdateVersionedTransition = currentVersionedTransition
+	}
+	for _, timerInfo := range ms.updateTimerInfos {
+		timerInfo.LastUpdateVersionedTransition = currentVersionedTransition
+	}
+	for _, childInfo := range ms.updateChildExecutionInfos {
+		childInfo.LastUpdateVersionedTransition = currentVersionedTransition
+	}
+	for _, cancelInfo := range ms.updateRequestCancelInfos {
+		cancelInfo.LastUpdateVersionedTransition = currentVersionedTransition
+	}
+	for _, signalInfo := range ms.updateSignalInfos {
+		signalInfo.LastUpdateVersionedTransition = currentVersionedTransition
+	}
```

Comment:
Yeah for HSM will we can update the field since we know any changes to the state count as a transition. For existing fields it might not be the case, so doing it when closing transaction make things much easier.